ecs-agent/api/ecs/client/ecs_client.go

```diff
@@ -0,0 +1,801 @@
+	}
+	return attributeMap
+}
+
+func findMissingAttributes(expectedAttributes, actualAttributes map[string]string) ([]string, error) {
+	missingAttributes := make([]string, 0)
+	var err error
+	for key, val := range expectedAttributes {
+		if actualAttributes[key] != val {
+			missingAttributes = append(missingAttributes, key)
+		} else {
+			logger.Trace("Response contained expected value for attribute", logger.Fields{
+				"key": key,
+			})
+		}
+	}
+	if len(missingAttributes) > 0 {
+		err = apierrors.NewAttributeError("Attribute validation failed")
+	}
+	return missingAttributes, err
+}
+
+func (client *ecsClient) getResources() ([]*ecsmodel.Resource, error) {
+	// Below are micro-optimizations - the pointers to integerStr and stringSetStr are used multiple times below.
+	integerStr := "INTEGER"
+	stringSetStr := "STRINGSET"
+
+	cpu, mem := getCpuAndMemory()
+	remainingMem := mem - int64(client.configAccessor.ReservedMemory())
+	logger.Info("Remaining memory", logger.Fields{
+		"remainingMemory": remainingMem,
+	})
+	if remainingMem < 0 {
+		return nil, fmt.Errorf(
+			"api register-container-instance: reserved memory is higher than available memory on the host, "+
+				"total memory: %d, reserved: %d", mem, client.configAccessor.ReservedMemory())
+	}
+
+	cpuResource := ecsmodel.Resource{
+		Name:         aws.String("CPU"),
+		Type:         &integerStr,
+		IntegerValue: &cpu,
+	}
+	memResource := ecsmodel.Resource{
+		Name:         aws.String("MEMORY"),
+		Type:         &integerStr,
+		IntegerValue: &remainingMem,
+	}
+	portResource := ecsmodel.Resource{
+		Name:           aws.String("PORTS"),
+		Type:           &stringSetStr,
+		StringSetValue: utils.Uint16SliceToStringSlice(client.configAccessor.ReservedPorts()),
+	}
+	udpPortResource := ecsmodel.Resource{
+		Name:           aws.String("PORTS_UDP"),
+		Type:           &stringSetStr,
+		StringSetValue: utils.Uint16SliceToStringSlice(client.configAccessor.ReservedPortsUDP()),
+	}
+
+	return []*ecsmodel.Resource{&cpuResource, &memResource, &portResource, &udpPortResource}, nil
+}
+
+// GetHostResources calling getHostResources to get a list of CPU, MEMORY, PORTS and PORTS_UPD resources
+// and return a resourceMap that map the resource name to each resource
+func (client *ecsClient) GetHostResources() (map[string]*ecsmodel.Resource, error) {
+	resources, err := client.getResources()
+	if err != nil {
+		return nil, err
+	}
+	resourceMap := make(map[string]*ecsmodel.Resource)
+	for _, resource := range resources {
+		if *resource.Name == "PORTS" {
+			// Except for RCI, TCP Ports are named as PORTS_TCP in Agent for Host Resources purpose.
+			resource.Name = aws.String("PORTS_TCP")
+		}
+		resourceMap[*resource.Name] = resource
+	}
+	return resourceMap, nil
+}
+
+func getCpuAndMemory() (int64, int64) {
+	memInfo, err := meminfo.Read()
+	mem := int64(0)
+	if err == nil {
+		mem = memInfo.MemTotal / 1024 / 1024 // MiB
```

Comment:
This comment may not be entirely accurate. Looking at the source for moby `meminfo.Read()`

https://github.com/moby/moby/blob/v24.0.7/pkg/meminfo/meminfo.go#L8-L10 Calls:

https://github.com/moby/moby/blob/master/pkg/meminfo/meminfo_linux.go#L38-L44

Which converts to and returns mem in KiB. It seems like this is returning GiB?

Reply:
Synced up offline. The memory being returned here is in MiB.

ref: https://access.redhat.com/documentation/en-us/red_hat_enterprise_linux/6/html/deployment_guide/s2-proc-meminfo

> MemTotal — Total amount of usable RAM, in kibibytes, which is physical RAM minus a number of reserved bits and the kernel binary code.